### PR TITLE
Fix 'command not found' errors under different circustamces

### PR DESCRIPTION
**Pull request by juan131**
**Description of the change**

The **PATH** environment variable is reseted under certain circumstances (e.g. when using `sudo` to execute commands as a different user, when accessing via SSH, etc.). This is highly problematic when you try to use Phabricator to host your repositories and access them via SSH because:

- You need to access the container via SSH using a **PHABRICATOR_SSH_VCS_USER** user.
- SSH daemon is configured to use Phabricator CLI to check authorised keys (requires customisations at `/etc/ssh/sshd`).
- The **PHABRICATOR_SSH_VCS_USER** user needs to perform some actions as the **PHABRICATOR_DAEMON_USER** user (requires `/etc/sudoers` customisations).
- Nor `php` nor `git` related commands are in default linux paths. 

After some investigation, I could get it working by:

- Adapting `secure_path ` in `/etc/sudoers`.
- Adding **PATH** to `env_keep` in `/etc/sudoers`.
- Using `PermitUserEnvironment yes` in `/etc/ssh/sshd_config`..
- Adding the **PATH** to `/home/git/.ssh/environment` and `/home/phabricator/.ssh/environment`.

**Benefits**

Users can actually use Phabricator with hosted repositories using SSH authentication.

**Possible drawbacks**

None

**Applicable issues**

- https://github.com/bitnami/bitnami-docker-phabricator/issues/166

### CODE REVIEW

File: 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh [hunk 1]
```diff
@@ -43,10 +43,15 @@ if am_i_root; then
...
-"${PHABRICATOR_SSH_VCS_USER}" ALL=("${PHABRICATOR_DAEMON_USER}") SETENV: NOPASSWD: $(command -v git), $(command -v git-upload-pack), $(command -v git-receive-pack), $(command -v ssh)
+${WEB_SERVER_DAEMON_USER} ALL=(${PHABRICATOR_DAEMON_USER}) SETENV: NOPASSWD: $(command -v git), $(command -v git-http-backend), $(command -v ssh)
+${PHABRICATOR_SSH_VCS_USER} ALL=(${PHABRICATOR_DAEMON_USER}) SETENV: NOPASSWD: $(command -v git), $(command -v git-upload-pack), $(command -v git-receive-pack), $(command -v ssh)
 EOF
     chmod 440 /etc/sudoers.d/phabricator
+    # SSH + sudo combination resets the environment and we're finding a lot of issues related to commands unable to find binaries
+    # TODO: find a better way to define the environment for daemon + vcs users combined with sudo actions.
+    for binary in "php" "git" "git-http-backend" "git-upload-pack" "git-receive-pack"; do
+        ln -s "$(command -v "$binary")" "/usr/bin/${binary}"
+    done
```
Comment (marcosbc) [inline 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:54]: I'm not convinced with this. In the Provisioner tool we're adding PATH to these files:

    /root/{.bashrc,.profile}
    /home/*/{.bashrc,.profile}
    /etc/skel/{.bashrc,.profile}

You could also try and define `secure_path` in sudoers with an appropriate value to include Git.

And finally, if none of those convince you, try adding `PATH` to `env_keep` option in sudoers.

**Review by marcosbc — CHANGES_REQUESTED** [review]

File: 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh (hunk 1 above)
Comment (javsalgar) [reply 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:54]: I think the provisioner approach could work in this case. I imagine that in the Phabricator VM we don't find these issues, right?

Comment (juan131) [reply 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:54]: @marcosbc the `secure_path` + `env_keep` made the trick.

File: 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh [hunk 2]
```diff
@@ -42,11 +42,23 @@ if am_i_root; then
...
     # Web Server & VCS users need to be able to sudo as the PH daemon user so they can interact with repositories
-    cat > /etc/sudoers.d/phabricator << EOF
-"${WEB_SERVER_DAEMON_USER}" ALL=("${PHABRICATOR_DAEMON_USER}") SETENV: NOPASSWD: $(command -v git), $(command -v git-http-backend), $(command -v ssh)
-"${PHABRICATOR_SSH_VCS_USER}" ALL=("${PHABRICATOR_DAEMON_USER}") SETENV: NOPASSWD: $(command -v git), $(command -v git-upload-pack), $(command -v git-receive-pack), $(command -v ssh)
+    sudoers_tmp=$(mktemp)
+    ph_sudoers_tmp=$(mktemp)
+    cat /etc/sudoers > "$sudoers_tmp"
+    replace_in_file "$sudoers_tmp" "^Defaults\s+env_reset$" "Defaults\tenv_reset\nDefaults\tenv_keep += PATH"
+    replace_in_file "$sudoers_tmp" "^Defaults\s+secure_path=\"(.*)\"$" "Defaults\tsecure_path=\"\1:$(command -v git | xargs dirname):$(command -v php | xargs dirname)\""
+    visudo -c -q -f "$sudoers_tmp" && cat "$sudoers_tmp" > /etc/sudoers
```
Comment (marcosbc) [inline 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:50]: This seems to be checking the format of the new sudoers file. If so, could you describe that in a comment? Same below

File: 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh [hunk 3]
```diff
@@ -42,11 +42,23 @@ if am_i_root; then
...
     debug_execute usermod -p NP "$PHABRICATOR_SSH_VCS_USER"
     # Web Server & VCS users need to be able to sudo as the PH daemon user so they can interact with repositories
-    cat > /etc/sudoers.d/phabricator << EOF
-"${WEB_SERVER_DAEMON_USER}" ALL=("${PHABRICATOR_DAEMON_USER}") SETENV: NOPASSWD: $(command -v git), $(command -v git-http-backend), $(command -v ssh)
-"${PHABRICATOR_SSH_VCS_USER}" ALL=("${PHABRICATOR_DAEMON_USER}") SETENV: NOPASSWD: $(command -v git), $(command -v git-upload-pack), $(command -v git-receive-pack), $(command -v ssh)
+    sudoers_tmp=$(mktemp)
+    ph_sudoers_tmp=$(mktemp)
+    cat /etc/sudoers > "$sudoers_tmp"
+    replace_in_file "$sudoers_tmp" "^Defaults\s+env_reset$" "Defaults\tenv_reset\nDefaults\tenv_keep += PATH"
+    replace_in_file "$sudoers_tmp" "^Defaults\s+secure_path=\"(.*)\"$" "Defaults\tsecure_path=\"\1:$(command -v git | xargs dirname):$(command -v php | xargs dirname)\""
```
Comment (marcosbc) [inline 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:49]: Could you describe what is happening here in a comment? Same below

File: 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh [hunk 4]
```diff
@@ -41,12 +41,27 @@ if am_i_root; then
     ensure_user_exists "$PHABRICATOR_SSH_VCS_USER" --group "$PHABRICATOR_SSH_VCS_GROUP" --home "/home/$PHABRICATOR_SSH_VCS_USER" --system
     # Unlock VCS user without password authentication
     debug_execute usermod -p NP "$PHABRICATOR_SSH_VCS_USER"
+    # Sudoers configuration syntax is critical. To avoid issues editing these files, we create
+    # temporary files to apply changes and check the syntax with 'visudo' before confirming them.
+    sudoers_tmp=$(mktemp)
+    ph_sudoers_tmp=$(mktemp)
+    cat /etc/sudoers > "$sudoers_tmp"
```
Comment (javsalgar) [inline 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:48]: Any reason for not using `cp`?

Comment (juan131) [reply 2021/debian-10/rootfs/opt/bitnami/scripts/phabricator/setup.sh:48]: Switching to `cp` after reading this: https://unix.stackexchange.com/questions/30295/cp-vs-cat-to-copy-a-file

**Review by marcosbc — APPROVED** [review]
LGTM!

**Pull request merged**